controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,136 @@
+	return changed, nil
+}
+
+// Get the space role objects from the templates specified in the given `spaceRoles`
+// Returns the objects indexed by kind and name, or an error if something wrong happened when processing the templates
+func (r *spaceRolesManager) getSpaceRoles(ns *corev1.Namespace, spaceRoles []toolchainv1alpha1.NSTemplateSetSpaceRole) ([]runtimeclient.Object, error) {
+	// store by kind and name
+	spaceRoleObjects := []runtimeclient.Object{}
+	for _, spaceRole := range spaceRoles {
+		tierTemplate, err := getTierTemplate(r.GetHostCluster, spaceRole.TemplateRef)
+		if err != nil {
+			return nil, err
+		}
+		for _, username := range spaceRole.Usernames {
+			objs, err := tierTemplate.process(r.Scheme, map[string]string{
+				Namespace: ns.Name,
+				Username:  username,
+			})
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to process space roles template '%s' for the user '%s' in namespace '%s'", spaceRole.TemplateRef, username, ns.Name)
+			}
+			spaceRoleObjects = append(spaceRoleObjects, objs...)
+		}
+	}
+	return spaceRoleObjects, nil
+}
+
+// looks-up the `toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey` annotation on the namespace to find the last-applied space roles
+// Returns the objects indexed by kind and name, or an error
+func (r *spaceRolesManager) getLastAppliedSpaceRoles(ns *corev1.Namespace) ([]runtimeclient.Object, error) {
+	// read annotation to see what was applied last time, so we can compare with the new SpaceRoles and remove all outdated resources (based on their kind/names)
+	spaceRoles := []toolchainv1alpha1.NSTemplateSetSpaceRole{}
+	if currentSpaceRolesAnnotation, exists := ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey]; exists && currentSpaceRolesAnnotation != "" {
+		if err := json.Unmarshal([]byte(currentSpaceRolesAnnotation), &spaceRoles); err != nil {
+			return nil, errors.Wrap(err, "unable to decode current space roles in annotation")
+		}
+	}
+	return r.getSpaceRoles(ns, spaceRoles)
+}
+
+// fetchNamespacesByOwner returns all current namespaces belonging to the given user
+// i.e., labeled with `"toolchain.dev.openshift.com/workspace":<workspace>`
+// (see https://github.com/codeready-toolchain/host-operator/blob/master/deploy/templates/nstemplatetiers/appstudio/ns_appstudio.yaml)
+func fetchNamespacesByWorkspace(cl runtimeclient.Client, workspace string) ([]corev1.Namespace, error) {
+	// fetch all namespace with owner=username label
+	namespaces := &corev1.NamespaceList{}
+	if err := cl.List(context.TODO(), namespaces, listByWorkspaceLabel(workspace)); err != nil {
+		return nil, err
+	}
+	return namespaces.Items, nil
+}
+
+// listByWorkspaceLabel returns runtimeclient.ListOption that filters by label toolchain.dev.openshift.com/workspace equal to the given workspace name
+func listByWorkspaceLabel(w string) runtimeclient.ListOption {
+	labels := map[string]string{toolchainv1alpha1.WorkspaceLabelKey: w}
+	return runtimeclient.MatchingLabels(labels)
+}
```

Comment:
as it was discussed in the call - don't list them by `workspace` label as SpaceRoles are not appstudio specific feature.

instead of that, you could use the already existing function `fetchNamespacesByOwner`

Reply:
done in 182a0cf